disnake/ext/commands/params.py

```diff
@@ -179,19 +196,84 @@ def _xt_to_xe(xe: Optional[float], xt: Optional[float], direction: float = 1) ->
         return None
 
 
-class Injection:
+class Injection(Generic[P, T_]):
+    """Represents a slash command injection.
+
+    .. versionadded:: 2.3
+
+    .. versionchanged:: 2.6
+        Added keyword-only argument ``autocompleters``.
+
+    Attributes
+    ----------
+    function: Callable
+        The underlying injection function.
+    autocompleters: Dict[:class:`str`, Callable]
+        A mapping of injection's option names to their respective autocompleters.
+    """
+
     _registered: ClassVar[Dict[Any, Injection]] = {}
 
     function: Callable
+    autocompleters: Dict[str, Callable]
 
-    def __init__(self, function: Callable) -> None:
+    def __init__(self, function: Callable, *, autocompleters: Dict[str, Callable] = None) -> None:
         self.function = function
+        self.autocompleters = autocompleters if autocompleters else {}
+
+    async def __call__(
+        self, inter: ApplicationCommandInteraction, *args: P.args, **kwargs: P.kwargs
```

Comment:
why don't we enforce that the interaction parameter is first or second? we probably should.

Reply:
Injections and commands use the same code for parsing and handling parameters, and injection callbacks often don't need an interaction. Enforcing it for one but not the other would complicate the code a bunch, not to mention that changing it now would also be a breaking change.

---
You could cheat the descriptor solution and make it easy by doing something like this:
```py
def __get__(self, obj, objtype):
    if obj is not None:
        self.cog = obj
    return self
```
It's pretty cursed, assumes that an injection object will always be bound to the same cog (which should be a valid assumption though), and isn't safe when instance and type calls are interleaved, ~~but it would work in most cases~~.

Going over it again, the "proper" descriptor solution isn't actually that difficult, it's essentially what `ext.tasks.Loop.__get__` and `__call__` do as well:
```diff
diff --git a/disnake/ext/commands/params.py b/disnake/ext/commands/params.py
index 859c1d87..24d1318c 100644
--- a/disnake/ext/commands/params.py
+++ b/disnake/ext/commands/params.py
@@ -220,21 +220,24 @@ class Injection(Generic[P, T_]):
     def __init__(self, function: Callable, *, autocompleters: Dict[str, Callable] = None) -> None:
         self.function = function
         self.autocompleters = autocompleters if autocompleters else {}
+        self._injected: Optional[Cog] = None
 
-    async def __call__(
-        self, inter: ApplicationCommandInteraction, *args: P.args, **kwargs: P.kwargs
-    ) -> T_:
+    def __get__(self, obj: Optional[Any], objtype: Type[Any]) -> Self:
+        if obj is None:
+            return self
+        copy = type(self)(function=self.function, autocompleters=self.autocompleters)
+        copy._injected = obj
+        setattr(obj, self.function.__name__, copy)
+        return copy
+
+    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> T_:
         """Calls the underlying function that the injection holds.
 
         .. versionadded:: 2.6
         """
-        cog = getattr(inter.application_command, "cog", None)  # haxxor
-        base_args = [cog, inter] if cog else [inter]
-
-        if asyncio.iscoroutinefunction(self.function):
-            return await self.function(*base_args, *args, **kwargs)
-        else:
-            return self.function(*base_args, *args, **kwargs)
+        if self._injected is not None:
+            args = (self._injected, *args)  # type: ignore
+        return self.function(*args, **kwargs)
 
     @classmethod
     def register(

```

So tl;dr, our options for this as far as I can tell are:
1. proper descriptor (see above, as well as the SO answer but that one's unnecessarily complex)
2. cheat descriptor (see further above, makes some assumptions but easy to implement)
3. no descriptor + no `__call__` (just makes it a breaking change)

Sorry for making the PR this complicated, couldn't really anticipate that :sweat_smile:

Reply:
You'd no longer be able to do this:
```py
@commands.register_injection
async def get_config(locale: str, theme: str) -> Config:
    ...

get_config("en-US", "dark")

# this would work, though
get_config.function("en-US", "dark")
```

Reply:
In that case we should implement 1.

Reply:
Thanks to shift, done in [ef192c25a242e95ced67fcde52bcffd158a23361](https://github.com/DisnakeDev/disnake/pull/675/commits/ef192c25a242e95ced67fcde52bcffd158a23361)